File: UVa/UglyNumbers/Ugly.java
import java.util.*;

public class Ugly {
  public static void main(String[] args) {
    boolean[] uglies = new boolean[20000000];
    uglies[1] = true;
    int counter = 0;
    ArrayList<Integer> uglyList = new ArrayList<Integer>(1500);
    uglyList.add(1);
    for (int i = 0; uglyList.size()<1500; i++) {
      int j = uglyList.get(i);
      System.out.println("Working with " + j);
      if (!uglyList.contains(j*2)) uglyList.add(j*2);
      if (!uglyList.contains(j*3)) uglyList.add(j*3);
      if (!uglyList.contains(j*5)) uglyList.add(j*5);
    }
    Collections.sort(uglyList);
    System.out.println(uglyList.get(1499));
  }

  public static boolean isUgly(int i, boolean[] uglies) {
    if (i == 1) return true;
    if (uglies[i]) return true;

    if (i % 2 == 0) {
      if (isUgly(i/2, uglies)) {
        uglies[i] = true;
        return true;
      }
    }

    if (i % 3 == 0) {
      if (isUgly(i/3, uglies)) {
        uglies[i] = true;
        return true;
      }
    }

    if (i % 5 == 0) {
      if (isUgly(i/5, uglies)) {
        uglies[i] = true;
        return true;
      }
    }

    return false;
  }
}

File: UVa/UglyNumbers/ugly.cpp
#include <iostream>

using namespace std;

bool isUgly(int, bool[]);
bool isUgly(int);

int main() {

  cout << "The 1500'th ugly number is 859963392." << endl;
  return 0;

  bool uglies[1000000];
  for (int i = 0; i < 1000000; i++) uglies[i] = false;

  uglies[1] = true;
  int counter = 0;
  int i = 1;
  while (counter < 1500) {
    if (isUgly(i)) counter++;
    if (counter == 1500) cout << i << endl;
    i++;
  }

  //cout << counter << " " << i << endl;

  return 0;
}

bool isUgly(int number) {
  if (number == 1) return true;
  if (number % 2 == 0) return isUgly(number/2);
  if (number % 3 == 0) return isUgly(number/3);
  if (number % 5 == 0) return isUgly(number/5);
  return false;
}

bool isUgly(int number, bool uglies[]) {
  if (number == 1) return true;

  if (number % 2 == 0) {
    if (uglies[number/2]) {
      uglies[number] = true;
      return true;
    }
  }

  if (number % 3 == 0) {
    if (uglies[number/3]) {
      uglies[number] = true;
      return true;
    }
  }

  if (number % 5 == 0) {
    if (uglies[number/5]) {
      uglies[number] = true;
      return true;
    }
  }

  return false;
}
